### PR TITLE
🔥 feat: Add Support for Removing Routes

### LINKED ISSUE #3098: 📝 [Proposal]: Allow removing registered route
### Feature Proposal Description

Following the addition of `RebuildTree` method #2769, which enabled dynamic request handling, I propose enhancing this functionality by allowing the removal of registered routes.

Currently, while we can dynamically add routes, the ability to remove them is limited. When adding a route with the same path, instead of updating the existing handler, a new one is added, which prevents the updated handler from being invoked. To address this issue, and to provide a way to remove routes that are no longer needed, we require a method to completely remove the route.

### Alignment with Express API

N/A

### HTTP RFC Standards Compliance

N/A

### API Stability

The proposed feature will involve the addition of a single method, `RemoveRoute`, which will be used to unregister specific routes based on the path and HTTP methods. This function will maintain the stability of the API by ensuring that the removal process is straightforward and does not interfere with other routes or the overall routing structure.

```go
func (app *App) RemoveRoute(path string, methods ...string) {
    // TODO:
}
 
```

### Feature Examples

```go
func addMyRoute() {
    // Remove route if it already exists
    s.app.RemoveRoute("/hello", fiber.MethodHead, fiber.MethodGet)

    // Add new route
    s.app.Get("/hello", func(c *fiber.Ctx) error {
        return c.SendString("hello")
    })

    // Rebuild the routing tree to reflect changes
    s.app.RebuildTree()
}

s.app.Get("/define", func(c *fiber.Ctx) error {
    before := s.app.HandlersCount()
    addMyRoute()
    after := s.app.HandlersCount()

    return c.JSON(map[string]any{
        "hcBefore": before,
        "hcAfter":  after,
    })
})
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have searched for existing issues that describe my proposal before opening this one.
- [X] I understand that a proposal that does not meet these guidelines may be closed without explanation.

### PR DESCRIPTION
**Pull request by ckoch786**
# Description

- Add method to Remove Routes
- Updated register method to prevent duplicate routes.

Fixes #3098

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Enhancement (improvement to existing features and functionality)
- [x] Documentation update (changes to documentation)

### CODE REVIEW

File: router.go [hunk 1]
```diff
@@ -375,6 +387,56 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
+
+		// Get unique HTTP method identifier
+		m := app.methodInt(method)
+		if m == -1 {
+			continue // Skip invalid HTTP methods
+		}
+
+		// Find the index of the route to remove
+		index := slices.IndexFunc(app.stack[m], func(r *Route) bool {
+			return r.Path == path
```
Comment (efectn) [inline router.go:421]: Can we also add another method called `RemoveRouteByName` to remove routes by their names?

Comment (ckoch786) [reply router.go:423]: RemoveRouteByName would remove all routes with the given name regardless of their path correct?
For instance if removing route named "test" would the both following routes be removed?
/api/test
/test

File: router.go [hunk 2]
```diff
@@ -357,6 +370,7 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
 			Method:   method,
 			Handlers: handlers,
 		}
+		// Should this use a mutex lock?
```
Comment (ckoch786) [inline router.go:373]: Question for you guys in the comment.

File: router.go [hunk 3]
```diff
@@ -375,6 +389,56 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
+	return slices.ContainsFunc(app.stack[app.methodInt(method)], func(r *Route) bool {
+		if r.path == pathRaw {
+			return true
+		}
+		return false
+	})
+}
+
+// RemoveRoute is used to remove a route from the stack
+// TODO should Remove Route only be called dynamically? Can you use it to remove the /define route, leaving the app without
```
Comment (ckoch786) [inline router.go:402]: Question for you guys in the comment.

File: router.go [hunk 4]
```diff
@@ -375,6 +389,56 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
+		if index == -1 {
+			continue // Route not found
+		}
+
+		// Remove route from tree stack
+		app.stack[m] = slices.Delete(app.stack[m], index, index+1)
+	}
+
+	// TODO THIS
+	app.routesRefreshed = true
```
Comment (ckoch786) [inline router.go:434]: Question for you guys in the comment.

File: router.go [hunk 5]
```diff
@@ -375,6 +389,56 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
+	// OR THIS
+	// // rebuild the tree if it has been built before
+	// if !app.routesRefreshed {
+	// 	app.buildTree()
+	// }
+}
```
[the remark this answers is not in the document]
Comment (ckoch786) [reply router.go:440]: Decided to go with the first route and updated the method comment.

File: router.go [hunk 6]
```diff
@@ -375,6 +383,55 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
 
+func (app *App) routeExists(method string, pathRaw string) bool {
+	return slices.ContainsFunc(app.stack[app.methodInt(method)], func(r *Route) bool {
+		return r.path == pathRaw
+	})
+}
+
+// RemoveRoute is used to remove a route from the stack
+// You should call RebuildTree after using this to ensure consistency of the tree
+func (app *App) RemoveRoute(path string, middlewareCount int, methods ...string) {
```
Comment (gaby) [inline router.go:394]: That `middlewareCount` param is very confusing. What's the purpose of it?

Comment (ckoch786) [reply router.go:394]: I agree, I do not like having to pass in the middlewareCount however.  The reason I believe this is needed is because the register method currently uses the length of handlers to increment the global routeCount.  

Here we can see that the handlers count should always be at least one. 
```go
func (app *App) register(methods []string, pathRaw string, group *Group, handler Handler, middleware ...Handler) {
	handlers := middleware
	if handler != nil {
		handlers = append(handlers, handler)
	}
```
^
|  more register method code between
v
```go
	// Increment global handler count
	atomic.AddUint32(&app.handlersCount, uint32(len(handlers))) //nolint:gosec // Not a concern
```

When removing a route we need to ensure that the app.handersCount is updated as well.  Since the user can pass in optional middleware this count can be greater than one for a given route.

```go 
func (app *App) deleteRoute(path string, methods []string, middlewareCount int) {
	app.mutex.Lock()
	defer app.mutex.Unlock()

	// No middleware passed in, increment to account for routes handler function
	if middlewareCount == 0 {
		middlewareCount++
	}

	//Decrement global handler count
	atomic.AddUint32(&app.handlersCount, ^uint32(middlewareCount-1)) //nolint:gosec // Not a concern
```

I have also thought of just passing in the registered middleware to RemoveRoute method and then passing the number of handlers into the deleteRoute method.  Not sure if that would be any less confusing.

I am open to suggestions.

Comment (ckoch786) [reply router.go:394]: Actually now that I am thinking about this more I think I should be able to get this from the existing route with that path and method.   I am starting to work on the RemoveRouteByName method that @efectn asked for so I will have to figure out how to get the counts from the existing routes since I will have to find all routes that contain the name of the route that needs to be deleted anyway.

File: router.go (hunk 1 above)
Comment (ckoch786) [reply router.go:422]: I see now that a Route has a Name :)

File: router.go (hunk 6 above)
Comment (ckoch786) [reply router.go:394]: I figured something better out.  Working on cleaning it up now.

File: router.go [hunk 7]
```diff
@@ -375,6 +390,47 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
+}
+
+// RemoveRoute is used to remove a route from the stack
+// You should call RebuildTree after using this to ensure consistency of the tree
+// TODO write tests for this that explicitally delete a route right after adding it
+func (app *App) RemoveRoute(path string, methods ...string) {
+	app.mutex.Lock()
+	defer app.mutex.Unlock()
+	app.deleteRoute(path, methods)
+}
```
[the remark this answers is not in the document]
Comment (ckoch786) [reply router.go]: Please add an example for a Test removing nested/parameterized routes

File: router.go (hunk 1 above)
Comment (ckoch786) [reply router.go]: @efectn Just added the new RemoveRouteByName method in my latest commit.

File: router.go (hunk 6 above)
Comment (ckoch786) [reply router.go]: @gaby Just cleaned up the call to deleteRoute in my latest commit.  I was able to do away with passing in any additional params to the method I solved it by doing this instead within the deleteRoute method.

```go
		// Find the index of the route to remove
		index := slices.IndexFunc(app.stack[m], matchFunc)
		if index == -1 {
			continue // Route not found
		}

		route := app.stack[m][index]

		// Decrement global handler count
		atomic.AddUint32(&app.handlersCount, ^uint32(len(route.Handlers)-1)) // nolint:gosec // Not a concern
		// Decrement global route position
		atomic.AddUint32(&app.routesCount, ^uint32(0))

```

File: router_test.go [hunk 8]
```diff
@@ -369,31 +373,224 @@ func Test_Router_NotFound_HTML_Inject(t *testing.T) {
...
+			routes = append(routes, routeMsg)
+		}
+	}
+
+	for _, route := range routes {
+		require.Equal(tb, expectedRoutesCount, strings.Count(route.handlers, " "))
+	}
+}
+
+func verifyThereAreNoRoutes(tb testing.TB, app *App) {
```
[the remark this answers is not in the document]
Comment (ckoch786) [reply router_test.go:429]: I almost always forget this, sigh.

**Comment by gaby** [comment]
@ckoch786 The method has to be added to the docs under `docs/`. Also mention this new method in the `whats_new.md` file. You can add it after the `RebuildTree` section introduced in #3074

**Comment by gaby** [comment]
@ckoch786 In the root of the repo we have a `Makefile` to ease local development. See here: https://github.com/gofiber/fiber#-development

You can run the tests locally and also the lint tests.

**Comment by ckoch786** [comment]
@gaby I cleaned up my branch and just pushed the docs changes.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# Usage:
```
Comment (gaby) [inline Dockerfile:1]: What is this doing here?

File: docs/api/app.md [hunk 10]
```diff
@@ -761,3 +761,56 @@ func main() {
...
 In this example, a new route is defined and then `RebuildTree()` is called to ensure the new route is registered and available.
+
+
+## RemoveRoute
+
+This method removes a route by path.  You must call the `RebuildTree()` method after the remove in to ensure the route is removed.
+
+```go title="Signature"
+func (app *App) RemoveRoute()
+```
```
[the remark this answers is not in the document]
Comment (gaby) [reply docs/api/app.md:772]: Apply this change

[the remark this answers is not in the document]
Comment (gaby) [reply docs/api/app.md:772]: Also remove the extra space at the beginning of the 2nd sentence.

File: docs/api/app.md [hunk 11]
```diff
@@ -761,3 +761,56 @@ func main() {
...
+
+This method removes a route by name
+```go title="Signature"
+func (app *App) RemoveRouteByName()
+```
```
[the remark this answers is not in the document]
Comment (gaby) [reply docs/api/app.md:777]: Apply this too

File: docs/api/app.md [hunk 12]
```diff
@@ -761,3 +761,56 @@ func main() {
...
+           app.RemoveRoute("/api/feature", MethodGet)
+           app.RebuildTree()
+           // Redefine route
+           app.Get("/api/feature", func(c Ctx) error {
+                   return c.SendString("Testing feature-a")
+           })
+
+           app.RebuildTree()
+           return c.SendStatus(http.StatusOK)
+    })
```
[the remark this answers is not in the document]
Comment (gaby) [reply docs/api/app.md:801]: ^Apply the above

File: router_test.go [hunk 13]
```diff
@@ -369,31 +373,267 @@ func Test_Router_NotFound_HTML_Inject(t *testing.T) {
 	require.Equal(t, "Cannot DELETE /does/not/exist&lt;script&gt;alert(&#39;foo&#39;);&lt;/script&gt;", string(c.Response.Body()))
 }
 
+func sendStatusOK(c Ctx) error {
```
Comment (gaby) [inline router_test.go:376]: Remove this helper, do this call in the test

**Review by gaby — CHANGES_REQUESTED** [review]

**Review by gaby — COMMENTED** [review]
Some comments

**Comment by gaby** [comment]
@ckoch786 Test and lint workflows are failing. You can test them locally using the `Makefile`.

File: docs/api/app.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ckoch786) [reply docs/api/app.md:801]: @gaby the router_test.go file is already part of the fiber package so the Ctx is already the fiber Ctx.  Also MethodGet is also in the fiber package and that is what is being used. In addition MethodGet and StatusOK are both defined in the fiber package source file constants.go.  So I did update my tests to use  the fiber constant StatusOK instead of the http.StatusOK constant.

When I made the update it importing github.com/gofiber/fiber after doing so and running `make tidy` and then `make test` I get the following errors:
# github.com/gofiber/fiber/v3 [github.com/gofiber/fiber/v3.test]
./router_test.go:378:9: c.SendStatus(http.StatusOK) (no value) used as value
./router_test.go:382:19: cannot use func(c *fiber.Ctx) error {…} (value of type func(c *"github.com/gofiber/fiber".Ctx) error) as Handler value in argument to app.Get
./router_test.go:383:35: cannot use sendStatusOK (value of type func(c *"github.com/gofiber/fiber".Ctx) error) as Handler value in argument to app.Get
./router_test.go:387:10: c.SendStatus(http.StatusOK) (no value) used as value
./router_test.go:450:28: cannot use func(c *fiber.Ctx) error {…} (value of type func(c *"github.com/gofiber/fiber".Ctx) error) as Handler value in argument to app.Get
./router_test.go:454:27: cannot use func(c *fiber.Ctx) error {…} (value of type func(c *"github.com/gofiber/fiber".Ctx) error) as Handler value in argument to app.Get
./router_test.go:455:11: c.SendString("Testing feature-a") (no value) used as value
./router_test.go:459:10: c.SendStatus(fiber.StatusOK) (no value) used as value
./router_test.go:461:28: cannot use func(c *fiber.Ctx) error {…} (value of type func(c *"github.com/gofiber/fiber".Ctx) error) as Handler value in argument to app.Get
./router_test.go:465:27: cannot use func(c *fiber.Ctx) error {…} (value of type func(c *"github.com/gofiber/fiber".Ctx) error) as Handler value in argument to app.Get
./router_test.go:465:27: too many errors
FAIL    github.com/gofiber/fiber/v3 [build failed]

File: Dockerfile (hunk 9 above)
Comment (ckoch786) [reply Dockerfile:1]: I am stuck in Windows currently so I created the Dockerfile so that I could get all the dependencies for running `make` to install properly.  I have more updates that make it more useful that I will be pushing soon that utilizes the build caching and I updated the comments to include how to utilize volumes and bind mounts with run to keep preserve all the downloaded packages between runs and to allow the linter to update the local code that is outside the docker container.

However if you do not want me to add it I can just remove it.

What is nice about using docker for running the make actions is that if it is used it gives you an isolated container from the rest of your machine so that you can get consistent results.  Also using docker makes it easier for others to work with the Makefile as all they need to do is install Docker and the rest will just work.

Comment (gaby) [reply Dockerfile:1]: You cna remove it, this could be added to the docs though. Or one of the other fiber repos

Comment (ckoch786) [reply Dockerfile:1]: Also if you choose to keep it I can create helper scripts for running docker build and docker run to make it easier to use .

**Comment by ckoch786** [comment]
@gaby I am using the the Makefile lint and test.  The lint is now passing locally however the tests are really hit or miss for me locally not sure what is going on.  Even when running on  a fresh checkout of  gofiber/fiber on main three of the tests fail for me.
Also some of the tests that fail when running with make test pass when I run them with `go test -run Test_` so not really sure how to figure out why they are failing.  Any thoughts?

Also I did get to the point where all the test passed after running a few with `go test -run` and then rerunning `make test` but now I am back to 651 of them failing, not sure what I am doing wrong.

File: docs/whats_new.md [hunk 14]
```diff
@@ -941,6 +941,14 @@ In this example, a new route is defined, and `RebuildTree()` is called to ensure
...
 Note: Use this method with caution. It is **not** thread-safe and can be very performance-intensive. Therefore, it should be used sparingly and primarily in development mode. It should not be invoke concurrently.
 
+## RemoveRoute
+
+- **RemoveRoute**: Removes route by path
+
+- **RemoveRouteByName**: Removes route by name
+
+For more details, refer to the [app documentation](./api/app.md#removeroute):
+
```
[the remark this answers is not in the document]
Comment (ckoch786) [reply docs/whats_new.md:951]: This is already documented in the docs/api/app.md file.

**Comment by ckoch786** [comment]
@gaby I am looking at the failing tests from the actions that ran against my branch.  I think I found the the changes that I made  that is probably causing them to fail.  Though it is weird that they do not always fail for me locally, not sure what that is about yet.

**Comment by gaby** [comment]
> @gaby I am looking at the failing tests from the actions that ran against my branch. I think I found the the changes that I made that is probably causing them to fail. Though it is weird that they do not always fail for me locally, not sure what that is about yet.

Are you running just the new tests or the full suite?

Use this `make test`. We use `-shuffle` to run them in random orders, it helps catch weird edge cases.

**Comment by gaby** [comment]
@ckoch786 Any progress on this?

File: router.go [hunk 15]
```diff
@@ -395,6 +402,72 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
 	}
 }
 
+func (app *App) routeExists(method, pathRaw string) bool {
```
Comment (gaby) [inline router.go:405]: Something is wrong with this method. It's causing over 200 unit-tests to fail now.

**Comment by gaby** [comment]
@ReneWerner87 @efectn I vote we close this. Spent 1hr and can't find the issue, all the tests are failing. Author is not responding

File: router.go [hunk 16]
```diff
@@ -395,11 +396,81 @@ func (app *App) register(methods []string, pathRaw string, group *Group, handler
...
+		method = utils.ToUpper(method)
+
+		// Get unique HTTP method identifier
+		m := app.methodInt(method)
+		if m == -1 {
+			continue // Skip invalid HTTP methods
+		}
+
+		// Find the index of the route to remove
+		index := slices.IndexFunc(app.stack[m], matchFunc)
```
Comment (efectn) [inline router.go:451]: I think here we also need to look at other indexes in stack which also has matchFunc return true, not only the first one. It is necessary when a middleware is reqistered at the same path. Otherwise, we can skip `use` routes, which seems better idea to me.


Code:

```go
server.Use(func(c fiber.Ctx) error {
		return c.Next()
	})
server.Post("/", func(c fiber.Ctx) error {
	var t Test
	if err := c.Bind().Form(&t); err != nil {
		return err
	}
	fmt.Println(t)
	return c.SendString("Hello, World!")
})

server.Post("/", func(c fiber.Ctx) error {
	return c.SendString("Hello, World!")
})
```
Stack Output:

```go
0 {<nil> / GET  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 1 true false false true}
1 {<nil> / HEAD  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 2 true false false true}
2 {<nil> / POST  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 3 true false false true}
2 {<nil> / POST  / [] [0x7636c0 0x7637a0] {[0xc0000a2360] [] 0 0} 10 false false false true}
3 {<nil> / PUT  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 4 true false false true}
4 {<nil> / DELETE  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 5 true false false true}
5 {<nil> / CONNECT  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 6 true false false true}
6 {<nil> / OPTIONS  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 7 true false false true}
7 {<nil> / TRACE  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 8 true false false true}
8 {<nil> / PATCH  / [] [0x763660] {[0xc0000a22a0] [] 0 0} 9 true false false true}
```

Comment (ReneWerner87) [reply router.go:451]: yes please remove all with the same path

Comment (gaby) [reply router.go:451]: ^ Agree

**Comment by ckoch786** [comment]
> @ckoch786 Any progress on this?

@gaby No, sorry. Having to set up a dev environment where I could get the test to run properly set me back. That, and after seeing the number of tests that were failing, I was thinking that I just need to rip out the duplicate route detection logic because getting all the tests to pass is unfortunately going to take more time than I have available at the moment. If that is fine, then I can commit some time over the next few days to get the tests that are for the new RemoveRoute methods, since those do not touch any existing code. I will also update the docs.
Also, I see that you have made some changes to my PR. Thank you for not letting it get super stale!

**Comment by gaby** [comment]
> > @ckoch786 Any progress on this?
> 
> @gaby No, sorry. Having to set up a dev environment where I could get the test to run properly set me back. That, and after seeing the number of tests that were failing, I was thinking that I just need to rip out the duplicate route detection logic because getting all the tests to pass is unfortunately going to take more time than I have available at the moment. If that is fine, then I can commit some time over the next few days to get the tests that are for the new RemoveRoute methods, since those do not touch any existing code. I will also update the docs. Also, I see that you have made some changes to my PR. Thank you for not letting it get super stale!

I believe that @efectn has taken over the changes to this after some internal discussions. So no worries 💪

**Comment by ckoch786** [comment]
> @ReneWerner87 @efectn I vote we close this. Spent 1hr and can't find the issue, all the tests are failing. Author is not responding

@gaby I appreciate you trying to find the issue. Now I'm even more wary of messing with the duplicate route detection, as I don't know enough about Fiber at the moment to safely make this change. I was hoping that eliminating the ability to duplicate routes would not conflict with anything. It's a good thing that the tests caught it because my manual testing,  did not turn up any issues.

File: docs/api/app.md [hunk 17]
```diff
@@ -761,3 +761,62 @@ func main() {
...
+import (
+    "log"
+
+    "github.com/gofiber/fiber/v3"
+)
+
+func main() {
+    app := fiber.New()
+
+    app.Get("/api/feature-a", func(c *fiber.Ctx) error {
```
Comment (ReneWerner87) [inline docs/api/app.md:797]:
```suggestion
    app.Get("/api/feature-a", func(c fiber.Ctx) error {
```

File: docs/api/app.md [hunk 18]
```diff
@@ -761,3 +761,62 @@ func main() {
...
+           app.RebuildTree()
+           // Redefine route
+           app.Get("/api/feature", func(c *fiber.Ctx) error {
+                   return c.SendString("Testing feature-a")
+           })
+
+           app.RebuildTree()
+           return c.SendStatus(fiber.StatusOK)
+    })
+    app.Get("/api/feature-b", func(c *fiber.Ctx) error {
```
Comment (ReneWerner87) [inline docs/api/app.md:808]:
```suggestion
    app.Get("/api/feature-b", func(c fiber.Ctx) error {
```

File: docs/api/app.md [hunk 19]
```diff
@@ -761,3 +761,62 @@ func main() {
...
+           // Redefine route
+           app.Get("/api/feature", func(c *fiber.Ctx) error {
+                   return c.SendString("Testing feature-b")
+           })
+
+           app.RebuildTree()
+           return c.SendStatus(fiber.StatusOK)
+    })
+
+    log.Fatal(app.Listen(":3000"))
```
Comment (ReneWerner87) [inline docs/api/app.md:820]: pls also test the example in an example test case to verify the function

File: router.go [hunk 20]
```diff
@@ -313,6 +314,92 @@ func (*App) copyRoute(route *Route) *Route {
...
+// RemoveRouteFunc is used to remove a route from the stack by a custom match function.
+// This only needs to be called to remove a route, route registration prevents duplicate routes.
+// You should call RebuildTree after using this to ensure consistency of the tree.
+// Note: The route.Path is original path, not the normalized path.
+func (app *App) RemoveRouteFunc(matchFunc func(r *Route) bool, methods ...string) {
+	app.deleteRoute(methods, matchFunc)
+}
+
+func (app *App) deleteRoute(methods []string, matchFunc func(r *Route) bool) {
+	if len(methods) == 0 {
```
Comment (ReneWerner87) [inline router.go:363]: since the methods are required for all parent methods
we should not make it optional

maybe something like 
```go
func (app *App) RemoveRouteFunc(matchFunc func(r *Route) bool, method string, methods ...string)
```

or 

```go
func (app *App) RemoveRouteFunc(matchFunc func(r *Route) bool, method string[])
```
is better

Comment (ReneWerner87) [reply router.go:363]: Or we use all registered Methods as default -> then its okay to have it optional

File: router.go [hunk 21]
```diff
@@ -313,6 +314,92 @@ func (*App) copyRoute(route *Route) *Route {
...
+			// Decrement global handler count. In middleware routes, only decrement once
+			if _, ok := removedUseRoutes[route.path]; (route.use && slices.Equal(methods, app.config.RequestMethods) && !ok) || !route.use {
+				if route.use {
+					removedUseRoutes[route.path] = struct{}{}
+				}
+
+				atomic.AddUint32(&app.handlersCount, ^uint32(len(route.Handlers)-1)) //nolint:gosec // Not a concern
+			}
+		}
+	}
```
Comment (ReneWerner87) [inline router.go:400]: we could also directly call the RebuildTree function , then the users will not forget it

**Review by ReneWerner87 — CHANGES_REQUESTED** [review]
Some small hints

File: router.go (hunk 21 above)
Comment (efectn) [reply router.go:400]: what do you think @gaby @sixcolors

File: router.go (hunk 20 above)
Comment (efectn) [reply router.go:363]: Yes i will use default methods when methods is empty

**Comment by gaby** [comment]
@coderabbitai full review

File: docs/api/app.md [hunk 22]
```diff
@@ -761,3 +761,65 @@ func main() {
...
+
+This method removes a route by name.
+If no methods are specified, the route will be removed for all HTTP methods defined in the app. To limit removal to specific methods, provide them as additional arguments.
+
+```go title="Signature"
+func (app *App) RemoveRouteByName(name string, methods ...string)
+```
+
+This method removes a route by function having `*Route` parameter.
+If no methods are specified, the route will be removed for all HTTP methods defined in the app. To limit removal to specific methods, provide them as additional arguments.
```
Comment (gaby) [inline docs/api/app.md:782]: This same paragraph is repeated 3 times within 100 lines

File: router.go (hunk 21 above)
Comment (gaby) [reply router.go:400]: If you want to remove more than 1 route, then Rebuild will be called multiple times

**Review by gaby — APPROVED** [review]
👍 LGTM, although I'd remove that duplicated paragraph from the docs

**Review by ReneWerner87 — APPROVED** [review]
